staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -150,6 +151,59 @@ var (
 		[]string{"requestKind"},
 	)
 
+	// low and high concurrency watermarks from the last few windows
+	concurrencyWatermarks = compbasemetrics.NewGaugeVec(
+		&compbasemetrics.GaugeOpts{
+			Name:           "apiserver_inflight_count_watermarks",
+			Help:           "Low and high watermark of mutating, readonly requests executing in recent windows.",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"requestKind", "mark", "lag"},
+	)
+
+	concurrencyAverage = compbasemetrics.NewGaugeVec(
+		&compbasemetrics.GaugeOpts{
+			Name:           "apiserver_inflight_count_average",
+			Help:           "Average count of mutating, readonly requests executing during recent windows.",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"requestKind"},
+	)
+
+	concurrencyStddev = compbasemetrics.NewGaugeVec(
+		&compbasemetrics.GaugeOpts{
+			Name:           "apiserver_inflight_count_stddev",
+			Help:           "Standard deviation of count of mutating, readonly requests executing during recent windows.",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"requestKind"},
+	)
+
+	inflightCountElapsedSecondses = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_inflight_count_seconds",
+			Help:           "Seconds over which count of mutating, readonly requests executing have been integrated.",
```

Comment:
This metric doesn't make sense to me, can you elaborate what you're trying to answer with this metric?

Reply:
Since the metric isn't measured at collection time, this is keeping track of the cumulative amount of time over which the metric has been measured.  One alternative would be to add timestamps to the integral metrics below, but that would break any integrations with the push gateway.

Reply:
@dashpole : I am not sure what you mean by "add timestamps to the integral metrics".  Also, is this what that amounts to?

Reply:
The prometheus text format supports specifying explicit timestamps to describe when a metric was collected.  The prometheus client (although I suspect our compbasemetrics library doesn't yet) has a [NewMetricWithTimestamp](https://godoc.org/github.com/prometheus/client_golang/prometheus#NewMetricWithTimestamp) function that wraps a Metric and adds an explicit timestamp.  The problem you are trying to solve with this additional time metric is that the scrape time differs from the collection time.  One alternative, which would simplify the promql you have to write, is to add the collection timestamp to the metric explicitly.  Another alternative would be to update the integral counters at scrape time.

The biggest downside i'm aware of with explicit timestamps is that they don't work with the prometheus push gateway.